### PR TITLE
AIFBR-12310: skill restore in active active setup

### PR DESCRIPTION
**Pull request by vaibhavs-uipath**
Tested back up in the primary cluster and restore in secondary, skill pods came up.
Prediction of skill.

### CODE REVIEW

File: activeactive/skills/skill-restore.sh [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  PORT_FWS_PROCESS_ID=$(netstat -tulpn | grep 80 | grep kubectl | awk -F "[ /]+" '{print $7}')
+  kill -9 $PORT_FWS_PROCESS_ID
+
+  echo -e "$green Tenant Secret creation complete. Restoring backup dir \n"
+  echo $default
+}
+
+function restore_skill() {
+
+  kubectl apply -f $backupDir
```
Comment (vivekjhauipath) [inline activeactive/skills/skill-restore.sh:63]: won't this cause issue due to un-ordering?

**Review by vivekjhauipath — APPROVED** [review]
LGTM if tested, doesn't ordering matter?

File: activeactive/skills/skill-restore.sh (hunk 1 above)
Comment (vaibhavs-uipath) [reply activeactive/skills/skill-restore.sh:63]: I tried with the same order skill pods are up. Will confirm once with pramod.

Comment (vaibhavs-uipath) [reply activeactive/skills/skill-restore.sh]: corrected the order, the skill was deploying earlier but the public URL of prediction was not working.

File: activeactive/skills/skill-backup.sh [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+uuidArray=(${skillId//-/ })
+uuid=${uuidArray[0]}
+#echo "uuid: $uuid"
+mkdir $uuid-skill-backup
+
+function backup_deployment() {
+
+  getDeployment=`kubectl get deployment -n uipath | grep $skillId`
+    if [ -z "$getDeployment" ]; then
+        echo "$red $(date) Skill id is invalid, Please check ... Exiting $default"
```
Comment (vnaren23) [inline activeactive/skills/skill-backup.sh:24]: `No deployment present with provided skill id, Please check ... Exiting`
seems like a better error message

File: activeactive/skills/skill-backup.sh [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+echo "$green $(date) Taking backup of Skill with Id: $skillId"
+
+uuidArray=(${skillId//-/ })
+uuid=${uuidArray[0]}
+#echo "uuid: $uuid"
+mkdir $uuid-skill-backup
+
+function backup_deployment() {
+
+  getDeployment=`kubectl get deployment -n uipath | grep $skillId`
```
Comment (vnaren23) [inline activeactive/skills/skill-backup.sh:22]: `grep -i $skillId`
`-i` to ignore case

File: activeactive/skills/skill-restore.sh [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+# Validate required modules exits in target setup
+function validate_setup() {
+  validate_dependency jq "jq --version"
+  echo -e "$(date) Successfully validated required dependencies"
+}
+
+
+function create_tenant_secret() {
```
Comment (vnaren23) [inline activeactive/skills/skill-restore.sh:38]: Do we need to always create secret? Cant we skip if it already exists?

File: activeactive/skills/skill-restore.sh [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  S2S_CLIENT_SECRET=$(kubectl -n uipath get secret identity-client-aifabric -o json | jq '.data."Aicenter.Recovery.S2S.ClientSecret"' | sed -e 's/^"//' -e 's/"$//' | base64 -d)
+
+  ACCESS_TOKEN=$(curl --location --request POST $IDENTITY_SERVER_ENDPOINT \
+  --header 'Content-Type: application/x-www-form-urlencoded' \
+  --data-urlencode 'client_id='$S2S_CLIENT_ID \
+  --data-urlencode 'client_secret='$S2S_CLIENT_SECRET \
+  --data-urlencode 'grant_type=client_credentials' \
+  --data-urlencode 'audience=AiFabricRecovery' | jq '.access_token' | sed -e 's/^"//' -e 's/"$//')
+
+  kubectl -n uipath port-forward service/ai-deployer-svc 80:80 --address 0.0.0.0 &
```
Comment (vnaren23) [inline activeactive/skills/skill-restore.sh:53]: Externalize the port to some variable and have default value as 80. This way for whatever reason if customer wants to change port number they can only change at one place.

File: activeactive/skills/README.md [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+            Enter skill id:
+             083a05ac-1c1e-4db5-850b-ca1d9d4e56cd
+* This will generate a dir with all the artifacts of that skill.
+* Copy the dir in secondary.
+* Run script skill-restore.sh and give path of the copied folder as input.
+  * Example: 
+             [root@server0 testadmin]# sh skill-restore.sh
+             Enter backup dir path:
+              083a05ac-skill-backup
+
```
Comment (vnaren23) [inline activeactive/skills/README.md:21]: Please add prerequistes to readme. 
1. Dependencies like `jq`
2. Min AICenter version (22.10)

Comment (vnaren23) [reply activeactive/skills/README.md:21]: Also mention it requires port forwarding to work.

Comment (vnaren23) [reply activeactive/skills/README.md:21]: Also seeing other commands like `curl`, `netstat`. Please add those to prerequisites as well.

File: activeactive/skills/skill-restore.sh [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+
+function create_tenant_secret() {
+
+  echo -e "$green Starting port forwarding for ai-deployer to create tenant secret \n"
+
+  IDENTITY_SERVER_ENDPOINT=$(kubectl -n uipath get deployment ai-app-deployment -o json | jq -r '.spec.template.spec.containers[0].env[] | select(.name == "IDENTITY_SERVER_ENDPOINT").value')'/connect/token'
+  S2S_CLIENT_ID=$(kubectl -n uipath get secret identity-client-aifabric -o json | jq '.data."Aicenter.Recovery.S2S.ClientId"' | sed -e 's/^"//' -e 's/"$//' | base64 -d)
+  S2S_CLIENT_SECRET=$(kubectl -n uipath get secret identity-client-aifabric -o json | jq '.data."Aicenter.Recovery.S2S.ClientSecret"' | sed -e 's/^"//' -e 's/"$//' | base64 -d)
```
Comment (vnaren23) [inline activeactive/skills/skill-restore.sh:44]: Add validations to any input which we auto populate. 
Minimum validation should be empty check. IDENTITY_SERVER_ENDPOINT, S2S_CLIENT_ID, S2S_CLIENT_SECRET and ACCESS_TOKEN are all candidates.

File: activeactive/skills/skill-restore.sh [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  --data-urlencode 'audience=AiFabricRecovery' | jq '.access_token' | sed -e 's/^"//' -e 's/"$//')
+
+  kubectl -n uipath port-forward service/ai-deployer-svc 80:80 --address 0.0.0.0 &
+  sleep 5
+  echo -e "$green Calling create tenant secret API \n"
+  curl --silent --fail -i -k --show-error -X POST 'http://localhost:80/ai-deployer/v1/system/namespace/recover' -H 'authorization: Bearer '"$ACCESS_TOKEN"
+  echo -e "$yellow Waiting for tenant secret creation to complete  \n"
+  sleep 5
+  PORT_FWS_PROCESS_ID=$(netstat -tulpn | grep 80 | grep kubectl | awk -F "[ /]+" '{print $7}')
+  kill -9 $PORT_FWS_PROCESS_ID
```
Comment (vnaren23) [inline activeactive/skills/skill-restore.sh:60]: Is there any other way to gracefully close port forwarding? `kill` might spook customers.

**Review by vnaren23 — COMMENTED** [review]
Minor changes

File: activeactive/skills/skill-restore.sh (hunk 4 above)
Comment (vaibhavs-uipath) [reply activeactive/skills/skill-restore.sh:40]: not sure how to check this in script

File: activeactive/skills/skill-restore.sh (hunk 7 above)
Comment (vaibhavs-uipath) [reply activeactive/skills/skill-restore.sh]: done

File: activeactive/skills/skill-restore.sh (hunk 4 above)
Comment (vnaren23) [reply activeactive/skills/skill-restore.sh:40]: Why are we hitting `ai-deployer/v1/system/namespace/recover` endpoint? To create something. Lets find that and check if it exists.

File: activeactive/skills/skill-restore.sh (hunk 8 above)
Comment (vaibhavs-uipath) [reply activeactive/skills/skill-restore.sh:75]: not sure

File: activeactive/skills/skill-restore.sh (hunk 4 above)
Comment (vaibhavs-uipath) [reply activeactive/skills/skill-restore.sh:40]: this api will fetch all provisioned tenants and then get secretName and creates the secret

File: activeactive/skills/skill-restore.sh (hunk 8 above)
Comment (vnaren23) [reply activeactive/skills/skill-restore.sh:75]: Lets explore, if we cant find anything we'll leave it. Check with abhijit.chaudhari@uipath.com once.

File: activeactive/skills/skill-restore.sh (hunk 4 above)
Comment (vaibhavs-uipath) [reply activeactive/skills/skill-restore.sh:40]: i think API handles this, @pramod-rao-uipath  please confirm. I see from ML skill listener you are calling same API

File: activeactive/skills/skill-restore.sh [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+  --data-urlencode 'client_id='$S2S_CLIENT_ID \
+  --data-urlencode 'client_secret='$S2S_CLIENT_SECRET \
+  --data-urlencode 'grant_type=client_credentials' \
+  --data-urlencode 'audience=AiFabricRecovery' | jq '.access_token' | sed -e 's/^"//' -e 's/"$//')
+  if [[ -z $ACCESS_TOKEN ]]; then
+    echo "$red $(date) ACCESS_TOKEN is invalid or missing, Please check ... Exiting $default"
+    exit 1
+  fi
+
+  kubectl -n uipath port-forward service/ai-deployer-svc $PORT_FRWD:$PORT_FRWD --address 0.0.0.0 &
```
Comment (vnaren23) [inline activeactive/skills/skill-restore.sh:68]: Isnt it `$PORT_FRWD:80` ?
Source port should not be changed.

File: activeactive/skills/skill-restore.sh [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  sleep 5
+  echo -e "$green Calling create tenant secret API \n"
+  curl --silent --fail -i -k --show-error -X POST 'http://localhost:80/ai-deployer/v1/system/namespace/recover' -H 'authorization: Bearer '"$ACCESS_TOKEN"
```
Comment (vnaren23) [inline activeactive/skills/skill-restore.sh:71]: Use `$PORT_FRWD` instead of 80 here.

File: activeactive/skills/skill-restore.sh [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+  echo -e "$yellow Waiting for tenant secret creation to complete  \n"
+  sleep 5
+  PORT_FWS_PROCESS_ID=$(netstat -tulpn | grep 80 | grep kubectl | awk -F "[ /]+" '{print $7}')
```
Comment (vnaren23) [inline activeactive/skills/skill-restore.sh:74]: Same here. Use `$PORT_FRWD` instead of 80

File: activeactive/skills/skill-restore.sh (hunk 8 above)
Comment (vaibhavs-uipath) [reply activeactive/skills/skill-restore.sh]: done

File: activeactive/skills/skill-restore.sh (hunk 4 above)
Comment (vaibhavs-uipath) [reply activeactive/skills/skill-restore.sh:40]: done

File: activeactive/skills/README.md [hunk 12]
```diff
@@ -0,0 +1,31 @@
+# Script for manual backup and restore of skills in OnPrem
+
+## Purpose
+With skills being present in more than one cluster in Active-Active setup, there is a need to manually backup and 
+restore skills when needed.
+
+## Requirements
+The Machine where script runs needs the following:
+* Minimum AIC version 22.10
+* jq, curl, netstat to be installed.
```
Comment (vnaren23) [inline activeactive/skills/README.md:10]: nit: remove netstat since we are not using it anymore.

**Pull request merged**